game-core/src/main/java/games/strategy/triplea/delegate/DiceRoll.java

```diff
@@ -997,14 +997,9 @@ public static void getSortedSupport(
       final GameData data,
       final boolean defence,
       final boolean allies) {
-    final Set<UnitSupportAttachment> rules =
-        UnitSupportAttachment.get(data)
-            .parallelStream()
-            .filter(usa -> (usa.getRoll() || usa.getStrength()))
-            .collect(Collectors.toSet());
     getSupport(
         unitsGivingTheSupport,
-        rules,
+        data.getUnitTypeList().getSupportRules(),
```

Comment:
The rational is to cache it on the per GameData instance. Originally I had it in GameData, but GameData was already doing a lot of different things and had an abstraction for its UnitTypeList, so it seems like a natural place to put it.

Reply:
This might be a damned if you do, damned if you don't kind of situation.

With the move, we're going to be assuming that `UnitTypeList` is a singleton, which is arguably a bit assumption. `UnitTypeList` is also not much of a data structure, I think it's a candidate for us to inline all of its methods to where they are used and removed.

`Support` rules and the existing support code totally do not belong in `DiceRoll` either. Though, reaching out to another class to grab data to be used in a local computation is also a code smell as well.

I think a third approach, extracting the support method and the support cache to a new class so that it does not live in either DiceRoll nor UnitTypeList is probably the way to go. I can see that being too much for the scope of this update.

I'd be most comfortable if we had the support rules stay local to `DiceRoll` for now, or going with the expanded scope route. 

One concern is for serialization, as a `GameDataComponent` the `UnitTypeList` will be serialized to disk. That means after this update, any save game will *require* the support set field to be present, moving it later will break save-game compatibility. I think that would perhaps be the biggest reasoning to avoid a cached object in `GameDataComponent`, so that transient data is not persisted to save game files.

Reply:
The cache has to be per-game data, though.

So we either need to hang it off GameData (does that mean it will be part of save games?) or we hold a map keyed by GameData which is extra complexity and either a memory leak or we need a way to clear that cache otherwise.

Reply:
Also, my current implementation doesn't actually assume UnitTypeList is a singleton.

If in the future, this is called on other UnitTypeLists, everything works fine. They cache their associated support rules and everything works fine.

Given that any solution where the cache doesn't hang off GameData somehow results in a memory leak or needing to do lifetime management (e.g. when a game is closed, to go clear out these caches that would be keyed by GameData), I don't think it's a good idea to keep the cache separate.

For save game serialization, we can mark the field transient. WDYT?

Reply:
Anything that extends `GameDataComponent` will become part of save games.

Making it transient could do the trick indeed. Just please be sure to double check that it nulls out when you save and load and that the cache is refreshed after loading.

Reply:
Thanks! I did the test and confirmed that loading the save game results in the field being null.

Reply:
Cool, thanks for checking :+1: